packages/util/src/object.ts

```diff
@@ -108,24 +110,103 @@ export const assoc = <T>(key: string, value: any, obj: T) => ({
-  const flatResult: any[] = [];
-  const stack: any[] = [...input];
+export function toSanitizedObject(
+  key: string,
+  value: any
+): Record<string, any> {
+  return isNullish(key) || isNullish(value) ? {} : { [key]: value };
+}
 
-  while (stack.length > 0) {
-    const current = stack.pop();
-    if (!Array.isArray(current)) {
-      flatResult.push(current);
-    } else {
-      stack.push(...current);
-    }
-  }
+/**
+ * Pick a partial object from the given object indepdendent of casing - if available.
+ * @param obj - Object to pick the given key from.
+ * @param keys - Keys of the pair to be picked.
+ * @returns - An object containing the given key-value pairs in its original case or an empty object if none of them are found.
+ */
+export function pickIgnoreCase<T extends Record<string, any>>(
+  obj: T = {} as T,
+  ...keys: string[]
+): Partial<Pick<T, typeof keys[number]>> {
+  return keys.reduce((filteredHeaders, providedKey) => {
+    const originalKey = Object.keys(obj).find(
+      objKey => objKey.toLowerCase() === providedKey.toLowerCase()
+    );
 
-  return flatResult.reverse();
-};
+    return {
+      ...filteredHeaders,
+      ...(originalKey && { [originalKey]: obj[originalKey] })
+    };
+  }, {});
+}
+
+/**
+ * Pick the value of an object based on the given key, independent of casing.
+ * @param obj - Object to be searched for the given key.
+ * @param key - Key of the value to pick.
+ * @returns The value of for the given key or undefined if not available.
+ */
+export function pickValueIgnoreCase<T extends Record<string, any>>(
+  obj: T = {} as T,
+  key: string
+): any | undefined {
+  return Object.values(pickIgnoreCase(obj, key))[0];
+}
+
+/**
+ * Pick entries that have non-nullish values (and keys).
+ * @param obj - An object to pick frmo.
+ * @returns - A filtered object containing only headers with non-nullish values.
+ */
+export function pickNonNullishIgnoreCase<T extends Record<string, any>>(
+  obj: T = {} as T
+): Partial<T> {
+  return Object.entries(obj)
+    .filter(([key, value]) => !isNullish(key) && !isNullish(value))
+    .reduce((filtered, [key, value]) => ({ ...filtered, [key]: value }), {});
+}
+
+/**
+ * Create an object by merging the `right` object into the `left` object ignoring casing, but keeping the right casing. Only keys present in the left object will be present in the merged object.
+ * @param left - Object to merge into. They keys of this object will be present in the returned object.
+ * @param right - Object to merge. Only keys in `left` will be considered for merging.
+ * @returns - An object containing all keys from the `left` object, where headers present in the `right` object are replaced. Note that the casing used by `right` will be used.
+ */
+export function mergeLeftIgnoreCase<
+  LeftT extends Record<string, any>,
+  RightT extends Record<string, any>
+>(
+  left: LeftT = {} as LeftT,
+  right: RightT = {} as RightT
+): Record<string, any> {
+  return Object.entries(left)
```

Comment:
This does the right thing but I find it hard to understand. I would have expect: 
- Init the result with the right (because right wins)
- Loop over keys of left and only add the left key value if the pick does not find something
- Perhaps you can even create an inverse to the `pick` like unpick(keys) so that with a single call you get the stuff left in left and can add it.
- Also to call the single key method so often is highly inefficient because in each of them you loop over all keys.

Reply:
The steps you describe would not lead to the correct result, because consider the following case:
left: `{ a: 1, b: 2 }`, right: `{ A: 3, C: 4 }` => expected result `{ A: 3, b: 2, C: 4 }`.

I currently don't see a more efficient way than to:
1. Iterate over keys of left.
2. If right has the key (ignoring case), add it.
3. If not, add the one from left.

So, I could put everything into the reduce directly. When writing it I thought it might be a little simpler to split it into a map reduce.

Reply:
init result: `{ A: 3, C: 4 }`
result pickInverse(keysRight,left) is `{b:2}`
merge the two leads the expected result.

You call the 
```
pickValueIgnoreCase(right, key)  //calls pickIgnoreCase
        ? pickIgnoreCase(right, key)
```

so you actually the method twice. But I think it is pre mature optimization and we can adjust the implementation if at some point in time we hear from customers.

Reply:
Sorry, the example I gave was wrong. 🤦‍♀️ In that case you are right it would be easier the way you are proposing.
The correct example is:
left: `{ a: 1, b: 2 }`, right: `{ A: 3, C: 4 }` => expected result `{ A: 3, b: 2 }`